accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,671 @@
+
+Other options discussed include "`*()`" and "`*(comma_for_each_parameter)`". Consider that generics do not provide parameter type information either:
+```cs
+    string s = typeof(Action<int>).Name; // "Action`1
+```
+
+### `Namespace` property
+Returns `null`.
+
+A standard pointer returns the namespace of the referenced type, but we don't have that with function pointers since they are declared inline.
+
+### `IsPointer` property
+Returns `false`. Per past design discussions, this should be `false` since function pointers and standard pointers are not the same, and this also is non-breaking for existing code that needs to determine if a type is a standard pointer type.
+
+### `IsValueType` and `IsClass` properties
+Use the same behavior as standard pointers which is `IsValueType == false` and `IsClass == true`.
+
+The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
+- Type
+  - Value types
+    - Built-in value types
+    - User-defined value types
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+## Overidden members for a modified type
+Below are the non-trivial behaviors overridden by a modified type.
+
+### `bool IsModifiedType { get; }`
+Returns `true`.
+
+### `Type UnderlyingSystemType { get; }`
+The underlying type, such as `typeof(Int32)`.
+
+### `ToString()`
+Returns `UnderlyingSystemType.ToString()`. We could return a string representation of the underlying type plus its custom modifiers, but that is not considered a high enough priority at this time and the grammar would be somewhat arbitrary.
+
+### `Name`, `FullName`, `AssemblyQualifiedName`, `Assembly` and `Namespace` properties
+A modified type is an internal class called "System.RuntimeModifiedType" class that derives from a "System.ModifiedType" which derives from "System.Type". So these methods just return the expected values there.
+
+### `GetFunctionPointerParameters()`
+This returns modified types for each function pointer parameter which allows the user to call `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()`.
+
+This is listed here since this method also works on unmodified types, although unmodified types don't allow `GetRequiredCustomModifiers()` or `GetOptionalCustomModifiers()` to be called (an `InvalidOperationException` is thrown).
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // Throws InvalidOperationException if IsFunctionPointer = false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
++       public virtual bool IsUnmanagedFunctionPointer { get; }
```

Comment:
Should this just be `IsManagedFunctionPointer`? (remove the "un")

Reply:
> Should this just be `IsManagedFunctionPointer`? (remove the "un")

I think it should be "unmanaged" for consistency with existing names. Unmanaged is used in all existing related names: `unmanaged` keyword in C#, `UnmanagedCallersOnlyAttribute`, `SignatureCallingConvention.Unmanaged`.